python/paddle/distributed/auto_parallel/high_level_api.py

```diff
@@ -0,0 +1,465 @@
+                .global_block()
+                .ops[layer._op_recorder.start : layer._op_recorder.end]
+            )
+        layer._op_recorder.ops = ops
+
+
+def get_layer_pp_info(mesh, num_hidden_layers, layer_index):
+    if "pp" in mesh.dim_names:
+        pp_degree = mesh.get_dim_size("pp")
+        layer_per_stage = math.ceil(num_hidden_layers / pp_degree)
+        return layer_index // layer_per_stage
+    else:
+        # return None, False
+        return None
+
+
+# mesh, config: input_spec
+def to_distributed(model, dataloader, optimizer, mesh, config):
+    paddle.distributed.init_parallel_env()
+
+    with_pp = True if "pp" in mesh.dim_names else False
+    with_sp = True if config.sequence_parallel else False
+
+    # # Data Parallel
+    # # step_0: shard dataloader
+    if with_pp:
+        first_stage_mesh = mesh.get_mesh_with_dim("pp", 0)
+        last_stage_mesh = mesh.get_mesh_with_dim("pp", 1)
+        loader = dist.shard_dataloader(
+            dataloader,
+            meshes=[first_stage_mesh, last_stage_mesh],
+            shard_dims="dp",
+        )
+    else:
+        loader = dist.shard_dataloader(
+            dataloader, meshes=[mesh], shard_dims="dp"
+        )
+
+    # Sharding Parallel
+    # # step_1: shard optimizer
+
+    # # step_2: register pre-hooks and post-hooks, thus recording corresponding static ops in following paddle.jit.to_static
+    for layer in model.sublayers():
+        pre_hook_helper = layer.register_forward_pre_hook(
+            record_program_ops_pre_hook
+        )
+        post_hook_helper = layer.register_forward_post_hook(
+            record_program_ops_post_hook
+        )
+        layer._op_recorder.hooks.append(pre_hook_helper)
+        layer._op_recorder.hooks.append(post_hook_helper)
+
+    # # step_3: call @to_static, get program, and corresponding static ops of each layer
+    # (1) with FLAGS_enable_pir_api=False, get program based on var and op, default to False
+    # (2) with FLAGS_enable_pir_api=True, get pir program
+    static_func = paddle.jit.to_static(
+        model.forward, input_spec=config.input_spec, full_graph=True
+    )
+    program = static_func.concrete_program.main_program
+
+    # # step_4: get the mapping [dynamic-layers : static ops]
+    op_to_id = {}
+    for idx, op in enumerate(program.global_block().ops):
+        op_to_id[op] = idx
+
+    ops_id_to_layer = {}
+    op_id_to_layer = {}
+    for layer in model.sublayers():
+        layer_ops = layer._op_recorder.ops
+        ops_id = []
+        for op in layer_ops:
+            assert op in op_to_id.keys(), f"{op.name} is not in program"
+            op_id = op_to_id[op]
+            op_id_to_layer[op_id] = layer
+            ops_id.append(op_id)
+        ops_id_to_layer[tuple(ops_id)] = layer
+
+    # # step_5: pattern recogincation
+    DECODER_LAYER_NAME = 'decoder_layer'
+    register_used_patterns(DECODER_LAYER_NAME)
+    results = match_all_patterns(program)
```

Comment:
logging can be added here and some key locations for easy debugging.

Reply:
I will add logging in next PR. Thanks